codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Encoder.java

```diff
@@ -116,27 +117,26 @@ private void encodeHeadersEnforceMaxHeaderListSize(int streamId, ByteBuf out, Ht
                                                        SensitivityDetector sensitivityDetector)
             throws Http2Exception {
         long headerSize = 0;
+        // To ensure we stay consistent with our peer check the size is valid before we potentially modify HPACK state.
         for (Map.Entry<CharSequence, CharSequence> header : headers) {
             CharSequence name = header.getKey();
             CharSequence value = header.getValue();
-            long currHeaderSize = HeaderField.sizeOf(name, value);
             // OK to increment now and check for bounds after because this value is limited to unsigned int and will not
             // overflow.
-            headerSize += currHeaderSize;
```

Comment:
@nmittler - An alternative approach would be to stick with a single iteration, but close the connection if the limit is exceeded. The current approach was chosen to preserve stream independence at the cost of an extra iteration.